### PR TITLE
Docker template

### PR DESCRIPTION
**Pull request by ruffsl**
Adding support for Dockerfile generation:
- Support for prioritizing template packages to be use in finding template resources
- And unit tests for newly added sub functions and total Dockerfile generation

### CODE REVIEW

File: ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em [hunk 1]
```diff
@@ -1,3 +1,4 @@
+@[if base_image]@
```
Comment (tfoote) [inline ros_buildfarm/templates/snippet/from_base_image.Dockerfile.em]: Is this logic reversed? This looks like a new tag which won't necessarily be set and the old behavior will not persist. And the else uses base_name and base_tag_name which are also new.

File: ros_buildfarm/templates/__init__.py [hunk 2]
```diff
@@ -72,14 +75,18 @@ def _escape_value(value):
...
 
 def _expand_template(template_name, **kwargs):
-    global template_basepath
+    """Expand template call for interpreter"""
+    template_name = os.path.join('templates', template_name)
+    template_package = _find_first_template(template_name, **kwargs)
+
     global interpreter
-    template_path = os.path.join(template_basepath, template_name)
+    template_path = pkg_resources.resource_filename(template_package, template_name)
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/__init__.py]: Doesn't relying on `pkg_resources` to locate the templates break using the package from source? If yes, I would rather not want to sacrifice that.

File: ros_buildfarm/templates/__init__.py [hunk 3]
```diff
@@ -6,18 +6,18 @@
...
 
-template_basepath = os.path.abspath(os.path.dirname(__file__))
-
 interpreter = None
 template_hooks = None
 
 
 def expand_template(template_name, data, options=None):
-    global template_basepath
+    """Expand template call API"""
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/__init__.py]: The added doc blocks don't follow PEP 257.

File: ros_buildfarm/templates/__init__.py (hunk 2 above)
Comment (ruffsl) [inline ros_buildfarm/templates/__init__.py]: You can still use this if you add the package source dir to the python package path.
This is how I am able to code, run, debug the package from the source dir in my git repo.
I think if one would want to use a python package, they'd likely add it to the PYTHONPATH, but I'm not sure how else this being package might be used.

**Comment by ruffsl** [comment]
Hows this look?

**Comment by ruffsl** [comment]
One thing I just noticed. If the unit test is used with `nosetest` without the `--nocapture` argument for not capturing the stdout (i.e. any stdout output will be printed immediately), I get a good amount of errors with the interpreter:

``` shell
(venv) ruffsl@eve:~/git/ros-infrastructure/ros_buildfarm$ nosetests
.EE
======================================================================
ERROR: Test _find_first_template function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ruffsl/git/ros-infrastructure/ros_buildfarm/test/dockerfile_generation_test.py", line 71, in setUp
    interpreter.shutdown()
nose.proxy.UnboundLocalError: local variable 'interpreter' referenced before assignment
-------------------- >> begin captured stdout << ---------------------
Error processing /home/ruffsl/git/ros-infrastructure/ros_buildfarm/test/images.yaml.em

--------------------- >> end captured stdout << ----------------------

======================================================================
ERROR: Test _find_first_wrappers function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ruffsl/git/ros-infrastructure/ros_buildfarm/test/dockerfile_generation_test.py", line 71, in setUp
    interpreter.shutdown()
nose.proxy.UnboundLocalError: local variable 'interpreter' referenced before assignment
-------------------- >> begin captured stdout << ---------------------
Error processing /home/ruffsl/git/ros-infrastructure/ros_buildfarm/test/images.yaml.em

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 3 tests in 0.044s

FAILED (errors=2)
Exception ignored in: <bound method Interpreter.__del__ of <empy pseudomodule/interpreter at 0x7f10c5b5be10>>
Traceback (most recent call last):
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2094, in __del__
    self.shutdown()
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2157, in shutdown
    self.finalize()
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2646, in finalize
    self.push()
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2200, in push
    sys.stdout.push(self)
AttributeError: '_io.TextIOWrapper' object has no attribute 'push'
Exception ignored in: <bound method Interpreter.__del__ of <empy pseudomodule/interpreter at 0x7f10c5aaf940>>
Traceback (most recent call last):
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2094, in __del__
    self.shutdown()
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2157, in shutdown
    self.finalize()
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2646, in finalize
    self.push()
  File "/home/ruffsl/ws/deploy_ros_buildfarm/venv/lib/python3.4/site-packages/em.py", line 2200, in push
    sys.stdout.push(self)
AttributeError: '_io.TextIOWrapper' object has no attribute 'push'
```

as apposed to this

``` shell
(venv) ruffsl@eve:~/git/ros-infrastructure/ros_buildfarm$ nosetests --nocapture
Generating Dockerfile '/tmp/test/ros_build/dockerfile_dir/ros_core/Dockerfile':
...
----------------------------------------------------------------------
Ran 3 tests in 0.047s

OK
```

File: ros_buildfarm/templates/__init__.py [hunk 4]
```diff
@@ -91,26 +98,68 @@ def _expand_template(template_name, **kwargs):
...
+    data = dict(data)
+
+    # Prepend templates folder to name
+    data['template_name'] = os.path.join('templates',  data['template_name'])
+
+    template_name = data['template_name']
+    dockerfile_dir = data['dockerfile_dir']
+
+    # Find first instance of tempate_name given order of template_packages
+    template_package = _find_first_template(**data)
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/__init__.py]: Depending on the name of a positional argument (`template_name`) is very fragile.

File: ros_buildfarm/templates/__init__.py (hunk 2 above)
Comment (dirk-thomas) [inline ros_buildfarm/templates/__init__.py]: You are right. I always had the impression that most of pkg_resources functionality is only available after `install` / `develop`. Might this now allow to distribute an egg?

Comment (tfoote) [inline ros_buildfarm/templates/__init__.py]: Quite possibly, it's supposed to. But from our experience with the fragility of setuptools we'll need to test it.

**Comment by tfoote** [comment]
The files opened weren't being closed cleanly. Fixed in this branch: https://github.com/ros-infrastructure/ros_buildfarm/tree/close_files

The top commit should be cherry picked/rebased on top of this. Also this now needs a rebase. 

I'm not sure what to do about unit tests failing without stdout. I tried different methods but couldn't isolate the problem.

**Comment by dirk-thomas** [comment]
Is this PR still something to be worked on and considered or should it be closed since being obsolete?

**Comment by ruffsl** [comment]
Yes, this has much of the template machinery used to auto generate the official ROS Docker Hub Dockerfiles. It seems to have grown though.

File: ros_buildfarm/templates/__init__.py (hunk 4 above)
Comment (ruffsl) [inline ros_buildfarm/templates/__init__.py]: What would be a more reliable substitute approach?

Comment (dirk-thomas) [inline ros_buildfarm/templates/__init__.py]: Passing `template_name` as the first argument and not putting it in the dictionary which is passed with `**`.

**Comment by dirk-thomas** [comment]
Is it intended to continue this at some point? If yes, we should keep this open. If not, I would suggest to close the PR.

**Comment by ruffsl** [comment]
Just followed the talk in #96 , but am not sure what direction was decided for searching templates and global passing of lists.

**Comment by dirk-thomas** [comment]
The referenced PR added a global `template_prefix_path` variable which defines the search path. This can be set to a custom value before evaluating templates to use arbitrary template locations.

@tfoote Please iterate with @ruffsl on this ticket if it should still be merged.

**Comment by dirk-thomas** [comment]
I will close this ticket for now. Please feel free to reopen it in the future.

**Comment by mikaelarguedas** [comment]
@tfoote to look at this when he has a few spare cycles

**Comment by tfoote** [comment]
I've been talking with @mikaelarguedas about how to reintegrate this with more automation of docker image builds. We'll catch up with @ruffsl when he's back in the office.

**Pull request closed without merging**